cmd/server-mux.go

```diff
@@ -134,16 +134,88 @@ func (c *ConnMux) Read(b []byte) (int, error) {
 	return c.Conn.Read(b)
 }
 
-// ListenerMux - encapuslates the standard net.Listener to inspect
+// ListenerMux wraps the standard net.Listener to inspect
 // the communication protocol upon network connection
+// ListenerMux also wraps net.Listener to ensure that once
+// Listener.Close returns, the underlying socket has been closed.
+//
+// - https://github.com/golang/go/issues/10527
+//
+// The default Listener returns from Close before the underlying
+// socket has been closed if another goroutine has an active
+// reference (e.g. is in Accept).
+//
+// The following sequence of events can happen:
+//
+// Goroutine 1 is running Accept, and is blocked, waiting for epoll
+//
+// Goroutine 2 calls Close. It sees an extra reference, and so cannot
+//  destroy the socket, but instead decrements a reference, marks the
+//  connection as closed and unblocks epoll.
+//
+// Goroutine 2 returns to the caller, makes a new connection.
+// The new connection is sent to the socket (since it hasn't been destroyed)
+//
+// Goroutine 1 returns from epoll, and accepts the new connection.
+//
+// To avoid accepting connections after Close, we block Goroutine 2
+// from returning from Close till Accept returns an error to the user.
 type ListenerMux struct {
 	net.Listener
 	config *tls.Config
+	// Cond is used to signal Close when there are no references to the listener.
+	cond *sync.Cond
+	refs int
```

Comment:
Locking is still needed in the way functionality can be provided using sync.Cond

This is what i am talking about  - https://golang.org/pkg/sync/#Cond.Wait

### sync#Cond.Wait

Wait atomically unlocks c.L and suspends execution of the calling goroutine. After later 
resuming execution, Wait locks c.L before returning. Unlike in other systems, Wait cannot 
return unless awoken by Broadcast or Signal.

Because c.L is not locked when Wait first resumes, the caller typically cannot assume 
that the condition is true when Wait returns. Instead, the caller should Wait in a loop:

```
c.L.Lock()
for !condition() {
    c.Wait()
}
... make use of condition ...
c.L.Unlock()
```